# Introduction

We have to learn some basic programming concepts before doing anything in Maching Learning or Data Science or any Projects 

# Python Basics

## Entering expressions

In [1]:
2+2

4

**Commonly used math operators**
- Addition 
- Subtraction
- Multiplication
- Division
- Integer Division
- Moduls/ Remainder
- Exponent/ Power

**Order of Precedence**

The order of operations (also called precedence) of Python math operators is similar to that of mathematics. The ** operator is evaluated first; the *, /, //, and % operators are evaluated next, from left to right; and the + and - operators are evaluated last (also from left to right)

In [15]:
2+3*6

20

In [16]:
(2 + 3) * 6

30

In [17]:
2**5

32

In [25]:
22/7

3.142857142857143

In [26]:
22//7

3

In [27]:
22 % 7

1

## Common data types

**Common Data Types**
- Integer - `int`
    - e.g. `0,50,-100`
- Floating-Point - `float`
    - e.g. `1.25, 3.14, -1.0, --0.5, 0.0`
- String - `str`
    - e.g. `'a', 'aa', 'aaa', 'Hello!', '11 cats'`

String can be enclosed in single or double quotes

## String concatenation & Replicaiton

The meaning of an operator (e.g. `+`) may change based on the data types of the values next to it


In [23]:
# Addition for int or float
1 + 2

3

In [22]:
# Concatenation for string data type
'Bruce' + 'Wayne'

'BruceWayne'

if you try to use the + operator on a string and an integer value, Python will not know how to handle this, and it will display an <font color = red>error message</font>.

In [29]:
'Bruce' + 42

TypeError: must be str, not int

The `*` operator when used on one string value and one integer value becomes string replication operator.

In [31]:
2*5

10

In [30]:
'Batman'*5

'BatmanBatmanBatmanBatmanBatman'

**Varaibles**
- It's like box in the computer’s memory where we can store values

**Variable Names** are case sensitive.<br>We  can name a variable anything as long as it obeys the following three rules:

- It can be only one word.
- It can use only letters, numbers, and the underscore `_` character.
- It can’t begin with a number.

e.g. of *invalid variable names*: 	
`current-balance, current balance 4account, total_$um, 'batman'`

## `Hello world` - First Python program

In [4]:
# This program says hello and asks for my name.

print('Hello world!')
print('What is your name?')    # ask for their name
myName = input()
print('It is good to meet you, ' + myName)
print('The length of your name is:')
print(len(myName))
print('What is your age?')    # ask for their age
myAge = input()
print (type(myAge))
print('You will be ' + str(int(myAge) + 1) + ' in a year.')

Hello world!
What is your name?
James Bond
It is good to meet you, James Bond
The length of your name is:
10
What is your age?
50
<class 'str'>
You will be 51 in a year.


### Program Breakdown

The following line is called a comment. Python ignores comments, and you can use them to write notes or remind yourself what the code is trying to do. Any text for the rest of the line following a hash mark `#` is part of a comment.

#### Comments

In [5]:
# This program says hello and asks for my name.

The `print()` function displays the `string` value inside the parentheses on the screen. Notice that the quotes are not printed to the screen. They just mark where the string begins and ends; they are not part of the string value.

#### `print()` function

In [6]:
print('Hello world!')
print('What is your name?')    # ask for their name

Hello world!
What is your name?


The `input()` function waits for the user to type some text on the keyboard and press ENTER key.
The user input will then be stored in `myName` variable

#### `input()` function

In [7]:
myName = input()

Gandolf the Gray


The expression `'It is good to meet you, ' + myName` is evaluated and is `print`ed on the screen.

In [8]:
print('It is good to meet you, ' + myName)

It is good to meet you, Gandolf the Gray


The `len()` function returns the number of characters in the given string.

#### `len()` function

In [9]:
print('The length of your name is:')
print(len(myName))

The length of your name is:
16


#### type() function

The `type()` function returns the data type of the value given

In [7]:
print('What is your age?')    # ask for their age
myAge = input()
print (type(myAge))           # the input functions usually returns a string data type

What is your age?
50
<class 'str'>


#### `str(), int(), float()` functions

These functions are used to change the data types of the values. <br>
As the `input()` function returns `str` data, `myAge` variable is of `str` data type. <br>
To add a numerical value to it, the value should be converted to `int`.<br> 
After performing the addition, the `int` value is converted back to `str` again so that it can be concatenated with other strings.

In [8]:
print('You will be ' + str(int(myAge) + 1) + ' in a year.')

50


# Flow control

In [9]:
import this

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!


In [2]:
import this

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!


In [7]:
print (this.s)

Gur Mra bs Clguba, ol Gvz Crgref

Ornhgvshy vf orggre guna htyl.
Rkcyvpvg vf orggre guna vzcyvpvg.
Fvzcyr vf orggre guna pbzcyrk.
Pbzcyrk vf orggre guna pbzcyvpngrq.
Syng vf orggre guna arfgrq.
Fcnefr vf orggre guna qrafr.
Ernqnovyvgl pbhagf.
Fcrpvny pnfrf nera'g fcrpvny rabhtu gb oernx gur ehyrf.
Nygubhtu cenpgvpnyvgl orngf chevgl.
Reebef fubhyq arire cnff fvyragyl.
Hayrff rkcyvpvgyl fvyraprq.
Va gur snpr bs nzovthvgl, ershfr gur grzcgngvba gb thrff.
Gurer fubhyq or bar-- naq cersrenoyl bayl bar --boivbhf jnl gb qb vg.
Nygubhtu gung jnl znl abg or boivbhf ng svefg hayrff lbh'er Qhgpu.
Abj vf orggre guna arire.
Nygubhtu arire vf bsgra orggre guna *evtug* abj.
Vs gur vzcyrzragngvba vf uneq gb rkcynva, vg'f n onq vqrn.
Vs gur vzcyrzragngvba vf rnfl gb rkcynva, vg znl or n tbbq vqrn.
Anzrfcnprf ner bar ubaxvat terng vqrn -- yrg'f qb zber bs gubfr!


In [5]:
print ("".join([this.d.get(c, c) for c in this.s]))

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!


In [6]:
import codecs
print (codecs.decode(this.s, 'rot13'))

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!
